handlers/auth.go

```diff
@@ -0,0 +1,112 @@
+package handlers
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/Daniorocket/cookit-backend/lib"
+	"github.com/Daniorocket/cookit-backend/models"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/crypto/bcrypt"
+)
+
+func (d *Handler) Login(w http.ResponseWriter, r *http.Request) {
+	login := &models.Login{}
+	err := json.NewDecoder(r.Body).Decode(login)
+	if err != nil {
+		log.Println("Error:", err)
+		if err := CreateApiResponse(w, nil, http.StatusBadRequest, "failed", "Failed to decode json"); err != nil {
+			log.Println("Error create Api response:", err)
+		}
+		return
+	}
+	passDB, err := models.GetPasswordByUsernameOrEmail(d.Client, d.DatabaseName, login.Username)
+	if err != nil {
+		log.Println("Error:", err)
+		if err := CreateApiResponse(w, nil, http.StatusBadRequest, "failed", "Invalid username or password"); err != nil {
+			log.Println("Error create Api response:", err)
+		} //User doesn't exists
+		return
+	}
+	if err := bcrypt.CompareHashAndPassword([]byte(passDB), []byte(login.Password)); err != nil {
+		log.Println("Error:", err)
+		if err := CreateApiResponse(w, nil, http.StatusBadRequest, "failed", "Invalid username or password"); err != nil {
+			log.Println("Error create Api response:", err)
+		}
+		return
+	}
+	expirationTime := time.Now().Add(15 * time.Minute)
+	tokenString, err := lib.CreateJWT(login.Username, expirationTime)
+	if err != nil {
+		CreateApiResponse(w, nil, http.StatusBadRequest, "failed", "Failed to create JWT")
+		return
+	}
+	CreateApiResponse(w, lib.Token{
+		Token:          tokenString,
+		ExpirationTime: expirationTime.UTC().String(),
+	}, http.StatusOK, "success", "none")
+}
+func (d *Handler) Register(w http.ResponseWriter, r *http.Request) {
+	cred := models.Credentials{}
+	if err := json.NewDecoder(r.Body).Decode(&cred); err != nil {
+		log.Println("Error:", err)
+		if err := CreateApiResponse(w, nil, http.StatusBadRequest, "failed", "Failed to decode json"); err != nil {
+			log.Println("Error create Api response:", err)
+		}
+		return
+	}
+	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(cred.Password), 8)
+	if err != nil {
+		log.Println("Error:", err)
+		if err := CreateApiResponse(w, nil, http.StatusBadRequest, "failed", "Failed to create user account"); err != nil {
+			log.Println("Error create Api response:", err)
+		}
+		return
+	}
+	user := &models.User{
+		ID:       uuid.NewV4().String(),
+		Email:    cred.Email,
+		Username: cred.Username,
+		Password: string(hashedPassword),
+	}
+	if err := models.RegisterUser(d.Client, d.DatabaseName, user); err != nil {
+		if err := CreateApiResponse(w, nil, http.StatusBadRequest, "failed", "Failed to create user account"); err != nil {
+			log.Println("Error create Api response:", err)
+		}
+		return
+	}
+	CreateApiResponse(w, nil, http.StatusOK, "success", "none")
+}
+func (d *Handler) Renew(w http.ResponseWriter, r *http.Request) {
+	tkn := r.Context().Value("token").(JWT)
```

Comment:
Pytanie z ciekawości, co się stanie jak `r.Context().Value("token")` nie będzie typu `(JWT)` i jak obsłużyć to w ładniejszy sposób? (Nie musisz tego zmieniać, pytam tylko żeby Cię sprawdzić :))

Reply:
Szukałem w Internecie jak można odczytać wartość z kontekstu i na coś takiego trafiłem(niestety tutaj jeśli dobrze rozumiem, to nie uzyskam "errora"). Problem jest właśnie mi znany - gdy jeszcze podczas pisania programu nie dodałem tego Handlera do autoryzacji, to wyskoczył mi błąd w kodzie, gdy próbowało tu odczytać token, a go w rzeczywistości nie było. Innym rozwiązaniem, które przychodzi mi do głowy, to wczytanie w ten sposób do obiektu tokena i puszczenie przez validatora, które jeszcze muszę dorobić w kodzie. Jeśli wymyślę coś innego to jeszcze tutaj odpiszę.

Reply:
https://stackoverflow.com/a/50940347/8030922